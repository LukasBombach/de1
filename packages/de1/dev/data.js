const service = {
  a001: "04 09 01 c1 51 09 2c 5e e0 00 00 00 00 00 00 00 00 00",
  a002: "00",
  a003: "00 00 00 00 00 00 00 00",
  a004:
    "00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00",
  a005: "00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00",
  a006: "",
  a007: "",
  a008: "",
  a009: "",
  a00a: "00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00",
  a00b: "00 a0 78 50 32 3c c8 5e 00",
  a00c: "00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00",
  a00d: "02 5c 00 44 00 00 21 ea 22 1e da 5a 00 5e 00 00 00 02 4d",
  a00e: "00 00",
  a00f: "01 03 01 00 60",
  a010: "02 60 16 bc b2 00 04 00",
  a011: "19 8b 05 00",
};

const eventData10seconds = [
  ["a00d", "2d 43 00 44 00 00 20 9e 21 0d 5e 5a 00 5e 00 00 00 02 49"],
  ["a00e", "00 00"],
  ["a00d", "2d 47 00 44 00 00 20 ba 21 0d d9 5a 00 5e 00 00 00 02 49"],
  ["a00e", "00 00"],
  ["a00d", "2d 50 00 44 00 00 20 f2 21 0e 81 5a 00 5e 00 00 00 02 49"],
  ["a011", "19 6b 05 00"],
  ["a00e", "00 00"],
  ["a00d", "2d 53 00 44 00 00 20 9e 21 0e 65 5a 00 5e 00 00 00 02 49"],
  ["a011", "19 6c 05 00"],
  ["a00d", "2d 6c 00 44 00 00 20 d6 21 0c 03 5a 00 5e 00 00 00 02 49"],
  ["a00d", "2d 85 00 44 00 00 20 c8 21 0d d3 5a 00 5e 00 00 00 02 49"],
  ["a011", "19 69 05 00"],
  ["a00d", "2d 9e 00 44 00 00 20 ba 21 09 dd 5a 00 5e 00 00 00 02 49"],
  ["a011", "19 6a 05 00"],
  ["a00d", "2d b7 00 44 00 00 20 ac 21 0e 7e 5a 00 5e 00 00 00 02 49"],
  ["a00d", "2d d0 00 44 00 00 20 ac 21 09 c9 5a 00 5e 00 00 00 02 49"],
  ["a00d", "2d e9 00 44 00 00 20 ac 21 0c d0 5a 00 5e 00 00 00 02 49"],
  ["a00d", "2e 02 00 44 00 00 20 9e 21 0f 67 5a 00 5e 00 00 00 02 49"],
  ["a011", "19 6b 05 00"],
  ["a00d", "2e 1b 00 44 00 00 20 9e 21 0d d9 5a 00 5e 00 00 00 02 49"],
  ["a011", "19 6a 05 00"],
  ["a00d", "2e 34 00 44 00 00 20 ba 21 06 28 5a 00 5e 00 00 00 02 49"],
  ["a011", "19 6b 05 00"],
  ["a00d", "2e 4d 00 44 00 00 20 ac 21 09 21 5a 00 5e 00 00 00 02 49"],
  ["a011", "19 6d 05 00"],
  ["a00d", "2e 66 00 44 00 00 20 c8 21 0a 65 5a 00 5e 00 00 00 02 49"],
];

const states = {
  water: { level: 25.6484375, startFillLevel: 5 },
  versions: {
    bluetooth: {
      apiVersion: 4,
      release: 9,
      commits: 449,
      changes: 81,
      sha: "e05e2c09",
    },
    firmware: { apiVersion: 0, release: 0, commits: 0, changes: 0, sha: "" },
  },
  shot: {
    timer: 46515,
    groupPressure: 0.0166015625,
    groupFlow: 0,
    mixTemp: 35.69140625,
    headTemp1: 36,
    headTemp2: 13,
    headTemp3: 178,
    setMixTemp: 90,
    setHeadTemp: 94,
    setGroupPressure: 0,
    setGroupFlow: 0,
    frameNumber: 2,
    steamTemp: 83,
    time: 1579198605437,
  },
  shotSettings: {
    steamSettings: 0,
    targetSteamTemp: 160,
    targetSteamLength: 120,
    targetHotWaterTemp: 80,
    targetHotWaterVol: 50,
    targetHotWaterLength: 60,
    targetEspressoVol: 200,
    targetGroupTemp: 94,
  },
  stateInfo: { state: "sleep", substate: "ready" },
  shotDescriptionHeader: {
    headerV: 1,
    numberOfFrames: 3,
    numberOfPreinfuseFrames: 1,
    minimumPressure: 0,
    maximumFlow: 6,
  },
  shotFrame: {
    frameToWrite: 2,
    flag: 96,
    setVal: 1.375,
    temp: 94,
    frameLen: 50,
    triggerVal: 0,
    maxVol: 0,
  },
  calibrate: {
    writeKey: "0",
    calCommand: 0,
    calTarget: 0,
    de1ReportedVal: 0,
    measuredVal: 0,
  },
};
