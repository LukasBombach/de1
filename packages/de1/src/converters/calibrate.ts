import { Converter } from "@sblendid/sblendid";
import Parser from "../parser";

export interface Calibrate {
  writeKey: string;
  calCommand: number;
  calTarget: number;
  de1ReportedVal: number;
  measuredVal: number;
}

function decode(data: Buffer): Calibrate {
  return new Parser<Calibrate>(data)
    .int("writeKey", v => v.toString(16))
    .char("calCommand")
    .char("calTarget")
    .int("de1ReportedVal", v => Math.round(100 * (v / 65536)) / 100)
    .intSigned("measuredVal", v => Math.round(100 * (v / 65536)) / 100)
    .vars();
}

const converter: Converter<Calibrate> = {
  uuid: "a012",
  decode,
};

export default converter;
