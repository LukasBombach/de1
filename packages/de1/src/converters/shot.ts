import { Converter } from "@sblendid/sblendid";
import Parser from "../parser";

export interface Shot {
  timer: number;
  groupPressure: number;
  groupFlow: number;
  mixTemp: number;
  headTemp1: number;
  headTemp2: number;
  headTemp3: number;
  setMixTemp: number;
  setHeadTemp: number;
  setGroupPressure: number;
  setGroupFlow: number;
  frameNumber: number;
  steamTemp: number;
  time: number;
}

export const herz = 50;

function decode(data: Buffer): Shot {
  const time = Date.now();
  const vars = new Parser<Shot>(data)
    .short("timer", v => Math.round(100 * (v / (herz * 2))))
    .short("groupPressure", v => v / 4096)
    .short("groupFlow", v => v / 4096)
    .short("mixTemp", v => v / 256)
    .char("headTemp1")
    .char("headTemp2")
    .char("headTemp3")
    .short("setMixTemp", v => v / 256)
    .short("setHeadTemp", v => v / 256)
    .char("setGroupPressure", v => v / 16)
    .char("setGroupFlow", v => v / 16)
    .char("frameNumber")
    .char("steamTemp")
    .vars();
  return { ...vars, time };
}

const converter: Converter<Shot> = {
  uuid: "a00d",
  decode,
};

export default converter;
