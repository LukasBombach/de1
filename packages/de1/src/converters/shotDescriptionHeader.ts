import { ValidConverter } from ".";
import Parser from "../parser";
import Serializer from "../serializer";

export interface ShotDescriptionHeader {
  headerV: number;
  numberOfFrames: number;
  numberOfPreinfuseFrames: number;
  minimumPressure: number;
  maximumFlow: number;
}

function decode(data: Buffer): ShotDescriptionHeader {
  return new Parser<ShotDescriptionHeader>(data)
    .char("headerV")
    .char("numberOfFrames")
    .char("numberOfPreinfuseFrames")
    .char("minimumPressure", v => v / 16)
    .char("maximumFlow", v => v / 16)
    .vars();
}

function encode(header: ShotDescriptionHeader): Buffer {
  return new Serializer()
    .char(header.headerV)
    .char(header.numberOfFrames)
    .char(header.numberOfPreinfuseFrames)
    .char(header.minimumPressure * 16)
    .char(header.maximumFlow * 16).buffer;
}

function validate(header: Partial<ShotDescriptionHeader>): boolean {
  return Serializer.validate(header, {
    headerV: Number(),
    numberOfFrames: Number(),
    numberOfPreinfuseFrames: Number(),
    minimumPressure: Number(),
    maximumFlow: Number(),
  });
}

const converter: ValidConverter<ShotDescriptionHeader> = {
  uuid: "a00f",
  decode,
  encode,
  validate,
};

export default converter;
