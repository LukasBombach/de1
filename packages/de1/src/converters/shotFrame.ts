import { Converter } from "sblendid";
import Parser from "../parser";

export interface ShotFrame {
  frameToWrite: number;
  flag: number;
  setVal: number;
  temp: number;
  frameLen: number;
  triggerVal: number;
  maxVol: number;
}

function decode(data: Buffer): ShotFrame {
  return new Parser<ShotFrame>(data)
    .char("frameToWrite")
    .char("flag")
    .char("setVal", v => v / 16)
    .char("temp", v => v / 2)
    .char("frameLen", convertF817ToFloat)
    .char("triggerVal", v => v / 16)
    .short("maxVol", convertBottom10OfU10P0)
    .vars();
}

function convertF817ToFloat(value: number): number {
  const highBit = value & 128;
  return highBit === 0 ? value / 10 : value & 127;
}

function convertBottom10OfU10P0(value: number): number {
  return value & 1023;
}

const converter: Converter<ShotFrame> = {
  uuid: "a010",
  decode
};

export default converter;
