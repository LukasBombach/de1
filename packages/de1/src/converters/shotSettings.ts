import Parser from "../parser";
import Serializer from "../serializer";

export interface ShotSettings {
  steamSettings: number;
  targetSteamTemp: number;
  targetSteamLength: number;
  targetHotWaterTemp: number;
  targetHotWaterVol: number;
  targetHotWaterLength: number;
  targetEspressoVol: number;
  targetGroupTemp: number;
}

const converter: Converter<ShotSettings> = {
  name: "shotSettings",
  uuid: "a00b",
  encode,
  decode
};

function decode(data: Buffer): ShotSettings {
  return new Parser<ShotSettings>(data)
    .char("steamSettings")
    .char("targetSteamTemp")
    .char("targetSteamLength")
    .char("targetHotWaterTemp")
    .char("targetHotWaterVol")
    .char("targetHotWaterLength")
    .char("targetEspressoVol")
    .short("targetGroupTemp", v => v / 256)
    .vars();
}

function encode(shotSettings: ShotSettings): Buffer {
  return new Serializer()
    .char(shotSettings.steamSettings)
    .char(shotSettings.targetSteamTemp)
    .char(shotSettings.targetSteamLength)
    .char(shotSettings.targetHotWaterTemp)
    .char(shotSettings.targetHotWaterVol)
    .char(shotSettings.targetHotWaterLength)
    .char(shotSettings.targetEspressoVol)
    .short(shotSettings.targetGroupTemp)
    .buffer();
}

export default converter;
