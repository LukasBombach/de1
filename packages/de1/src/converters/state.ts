import { Converter } from "sblendid";
import Parser from "../parser";
import Serializer from "../serializer";

export type State =
  | "sleep"
  | "goingToSleep"
  | "idle"
  | "busy"
  | "espresso"
  | "steam"
  | "hotWater"
  | "shortCal"
  | "selfTest"
  | "longCal"
  | "descale"
  | "fatalError"
  | "init"
  | "noRequest"
  | "skipToNext"
  | "hotWaterRinse"
  | "steamRinse"
  | "refill"
  | "clean"
  | "inBootLoader"
  | "airPurge";

// type States = { [S in State]: number };
type States = Record<State, number>;

const converter: Converter<State> = {
  // name: "state",
  uuid: "a002",
  encode,
  decode
};

export const states: States = {
  sleep: 0x00,
  goingToSleep: 0x01,
  idle: 0x02,
  busy: 0x03,
  espresso: 0x04,
  steam: 0x05,
  hotWater: 0x06,
  shortCal: 0x07,
  selfTest: 0x08,
  longCal: 0x09,
  descale: 0x0a,
  fatalError: 0x0b,
  init: 0x0c,
  noRequest: 0x0d,
  skipToNext: 0x0e,
  hotWaterRinse: 0x0f,
  steamRinse: 0x10,
  refill: 0x11,
  clean: 0x12,
  inBootLoader: 0x13,
  airPurge: 0x14
};

function decode(data: Buffer): State {
  const value = parse(data);
  const state = getNameFromValue(value);
  if (!state) throw new Error(`Received unexpected state ${value}`);
  return state;
}

function encode(state: State): Buffer {
  const value = states[state];
  if (value === undefined) throw new Error(`Unknown state "${state}"`);
  return serialize(state);
}

function parse(data: Buffer): number {
  return new Parser<{ state: number }>(data).char("state").vars().state;
}

function serialize(state: State): Buffer {
  return new Serializer().char(states[state]).buffer();
}

function getNameFromValue(state: number): State {
  const values = Object.values(states);
  const keys = Object.keys(states);
  const valueIndex = values.indexOf(state);
  if (valueIndex === -1) throw new Error(`Invalid state value "${state}"`);
  return keys[valueIndex] as State;
}

export default converter;
