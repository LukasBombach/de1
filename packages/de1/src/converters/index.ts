import { Converter } from "sblendid";

import state from "./state";
import water from "./water";
import versions from "./versions";
import shot from "./shot";
import shotSettings from "./shotSettings";
import stateInfo from "./stateInfo";

export type De1Converters = Converter<any>[];

const converters: De1Converters = [
  state,
  water,
  versions,
  shot,
  shotSettings,
  stateInfo
];

export default converters;
