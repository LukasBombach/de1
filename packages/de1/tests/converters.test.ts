import converters, { Converters } from "../src/converters";
import data from "./__fixtures__/characterics";

describe("converters", () => {
  test.each`
    name                       | uuid
    ${"calibrate"}             | ${"a012"}
    ${"shot"}                  | ${"a00d"}
    ${"shotDescriptionHeader"} | ${"a00f"}
    ${"shotFrame"}             | ${"a010"}
    ${"shotSettings"}          | ${"a00b"}
    ${"state"}                 | ${"a002"}
    ${"stateInfo"}             | ${"a00e"}
    ${"versions"}              | ${"a001"}
    ${"water"}                 | ${"a011"}
  `(
    "$name should have the uuid $uuid",
    ({ name, uuid }: { name: keyof Converters; uuid: string }) => {
      expect(uuid).toBe(converters[name].uuid);
    },
  );

  test.each(Object.entries(converters))(
    "decoding %s returns the expected output",
    (name, { uuid, decode }) => {
      if (decode) expect(decode(data[uuid])).toMatchSnapshot();
    },
  );

  test.each(Object.entries(converters).filter(([, { encode }]) => !!encode))(
    "encoding %s returns the expected output",
    (name, { uuid, decode, encode }) => {
      if (!decode) throw new Error("This test does not work");
      const decodedData = decode(data[uuid]);
      expect(encode!(decodedData)).toStrictEqual(data[uuid]);
    },
  );
});
