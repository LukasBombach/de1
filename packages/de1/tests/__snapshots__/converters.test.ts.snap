// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`converters decoding calibrate returns the expected output 1`] = `
Object {
  "calCommand": 0,
  "calTarget": 0,
  "de1ReportedVal": 0,
  "measuredVal": 0,
  "writeKey": "0",
}
`;

exports[`converters decoding shot returns the expected output 1`] = `
Object {
  "frameNumber": 3,
  "groupFlow": 0,
  "groupPressure": 0.01806640625,
  "headTemp": 75.57759094238281,
  "mixTemp": 75.0546875,
  "setGroupFlow": 3,
  "setGroupPressure": 0,
  "setHeadTemp": 89,
  "setMixTemp": 80,
  "steamTemp": 150,
  "timer": 7017,
}
`;

exports[`converters decoding shotDescriptionHeader returns the expected output 1`] = `
Object {
  "headerV": 1,
  "maximumFlow": 6,
  "minimumPressure": 0,
  "numberOfFrames": 6,
  "numberOfPreinfuseFrames": 1,
}
`;

exports[`converters decoding shotFrame returns the expected output 1`] = `
Object {
  "flag": 65,
  "frameLen": 127,
  "frameToWrite": 5,
  "maxVol": 0,
  "setVal": 1.8125,
  "temp": 88,
  "triggerVal": 0,
}
`;

exports[`converters decoding shotSettings returns the expected output 1`] = `
Object {
  "steamSettings": 0,
  "targetEspressoVol": 200,
  "targetGroupTemp": 89,
  "targetHotWaterLength": 60,
  "targetHotWaterTemp": 80,
  "targetHotWaterVol": 100,
  "targetSteamLength": 60,
  "targetSteamTemp": 160,
}
`;

exports[`converters decoding state returns the expected output 1`] = `"sleep"`;

exports[`converters decoding stateInfo returns the expected output 1`] = `
Object {
  "state": "sleep",
  "substate": "ready",
}
`;

exports[`converters decoding versions returns the expected output 1`] = `
Object {
  "bluetooth": Object {
    "apiVersion": 4,
    "changes": 81,
    "commits": 449,
    "release": 9,
    "sha": "92c5ee0",
  },
  "firmware": Object {
    "apiVersion": 0,
    "changes": 0,
    "commits": 0,
    "release": 0,
    "sha": "",
  },
}
`;

exports[`converters decoding water returns the expected output 1`] = `
Object {
  "level": 16.98828125,
  "startFillLevel": 6,
}
`;
